src/hotspot/share/prims/jvmtiEnv.cpp

```diff
@@ -988,38 +984,32 @@ JvmtiEnv::SuspendThreadList(jint request_count, const jthread* request_list, jvm
       results[i] = JVMTI_ERROR_NONE;  // indicate successful suspend
       continue;
     }
-
-    {
-      MutexLocker ml(java_thread->SR_lock(), Mutex::_no_safepoint_check_flag);
-      if (java_thread->is_external_suspend()) {
-        // don't allow nested external suspend requests.
-        results[i] = JVMTI_ERROR_THREAD_SUSPENDED;
-        continue;
-      }
-      if (java_thread->is_exiting()) { // thread is in the process of exiting
-        results[i] = JVMTI_ERROR_THREAD_NOT_ALIVE;
-        continue;
-      }
-      java_thread->set_external_suspend();
-    }
-    if (java_thread->thread_state() == _thread_in_native) {
-      // We need to try and suspend native threads here. Threads in
-      // other states will self-suspend on their next transition.
-      if (!JvmtiSuspendControl::suspend(java_thread)) {
-        // The thread was in the process of exiting. Force another
-        // safepoint to make sure that this thread transitions.
-        needSafepoint++;
+    if (java_thread->is_suspended()) {
+      results[i] = JVMTI_ERROR_THREAD_SUSPENDED;
+      continue;
+    }
+    if (java_thread == current) {
+      self_index = i;
+      continue;
+    }
+    if (!JvmtiSuspendControl::suspend(java_thread)) {
+      // Either the thread is already suspended or
+      // the thread was in the process of exiting:
+      if (java_thread->is_exiting()) {
         results[i] = JVMTI_ERROR_THREAD_NOT_ALIVE;
         continue;
       }
-    } else {
-      needSafepoint++;
+      results[i] =  JVMTI_ERROR_THREAD_SUSPENDED;
+      continue;
     }
     results[i] = JVMTI_ERROR_NONE;  // indicate successful suspend
   }
-  if (needSafepoint > 0) {
-    VM_ThreadsSuspendJVMTI tsj;
-    VMThread::execute(&tsj);
+  if (self_index >= 0) {
+    if (!JvmtiSuspendControl::suspend(current)) {
+      results[self_index] = JVMTI_ERROR_THREAD_NOT_ALIVE;
```

Comment:
My comment was about JVMTI_ERROR_THREAD_NOT_ALIVE

Reply:
It depends on how late we can call into an event handler during the
JavaThread exit code path. Is it possible to post an event on a thread
that has passed the `terminated` point in JavaThread::exit()?

Reply:
I think we should add JVMTI_ERROR_THREAD_SUSPENDED as @reinrich says, it is possible for someone to sneak in a second suspend request before us.

@dcubed-ojdk it seem like we could be posting JvmtiExport::post_monitor_contended_enter() from the ensure_join() which locks the threadObj.

So it might be best to treat this the same way as the others?

Reply:
By "treat this the same way as the others", you mean check and return either
JVMTI_ERROR_THREAD_NOT_ALIVE or JVMTI_ERROR_THREAD_SUSPENDED as
appropriate when we get a false back from JvmtiSuspendControl::suspend(current)?

I'm not sure what this question is about:

> it seem like we could be posting JvmtiExport::post_monitor_contended_enter() from the ensure_join() which locks the threadObj.

Reply:
I'm also unclear what Robbin is referring to. I go back to my original comment that surely JVMTI_ERROR_THREAD_NOT_ALIVE is impossible here?

Reply:
We do a callback with a terminated thread, if the thread then in the agent tries to suspend itself we should return JVMTI_ERROR_THREAD_NOT_ALIVE.

Reply:
"How can you do a callback on a terminated thread here? The thread should
only respond to the suspend request before it reaches the point in its
exit where it would report "not alive"."

We need to keep the threadObj while we are not terminated.
When terminated we need to clear threadObj and notify any waiters.
ObjectLocker executes OM::enter(), on contention OM::enter() do a callback to your agent.
If you on that callback execute suspendThreadList(), the correct return value for:
`"JvmtiSuspendControl::suspend(current)"` would be JVMTI_ERROR_THREAD_NOT_ALIVE, since JavaThread::is_exiting() would return true here.